File: ios/Sasquatsh/Services/EventsService.swift
import Foundation

protocol EventsServiceProtocol: Sendable {
    func getPublicEvents(filter: EventSearchFilter) async throws -> [EventSummary]
    func getRegisteredEvents() async throws -> [EventSummary]
    func getHostedEvents() async throws -> [EventSummary]
    func getEvent(id: String) async throws -> Event
    func createEvent(input: CreateEventInput) async throws -> Event
    func updateEvent(id: String, input: UpdateEventInput) async throws -> Event
    func deleteEvent(id: String) async throws
    func registerForEvent(eventId: String) async throws
    func cancelRegistration(eventId: String) async throws
    func addItem(eventId: String, input: CreateEventItemInput) async throws -> EventItem
    func claimItem(itemId: String) async throws
    func unclaimItem(itemId: String) async throws
    func addGame(input: AddEventGameInput) async throws -> EventGame
    func removeGame(gameId: String) async throws
}

final class EventsService: EventsServiceProtocol {
    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func getPublicEvents(filter: EventSearchFilter = EventSearchFilter()) async throws -> [EventSummary] {
        var items = filter.queryItems
        items.append(.init(name: "type", value: "browse"))
        return try await api.get("events", queryItems: items, authenticated: true)
    }

    func getRegisteredEvents() async throws -> [EventSummary] {
        try await api.get("events", queryItems: [.init(name: "type", value: "registered")], authenticated: true)
    }

    func getHostedEvents() async throws -> [EventSummary] {
        try await api.get("events", queryItems: [.init(name: "type", value: "hosted")], authenticated: true)
    }

    func getEvent(id: String) async throws -> Event {
        try await api.get("events", queryItems: [.init(name: "id", value: id)], authenticated: true)
    }

    func createEvent(input: CreateEventInput) async throws -> Event {
        try await api.post("events", body: input)
    }

    func updateEvent(id: String, input: UpdateEventInput) async throws -> Event {
        try await api.put("events", body: input, queryItems: [.init(name: "id", value: id)])
    }

    func deleteEvent(id: String) async throws {
        try await api.deleteVoid("events", queryItems: [.init(name: "id", value: id)])
    }

    func registerForEvent(eventId: String) async throws {
        let body = ["eventId": eventId]
        try await api.postVoid("registrations", body: body)
    }

    func cancelRegistration(eventId: String) async throws {
        try await api.deleteVoid("registrations", queryItems: [.init(name: "eventId", value: eventId)])
    }

    func addItem(eventId: String, input: CreateEventItemInput) async throws -> EventItem {
        struct ItemRequest: Codable {
            let eventId: String
            let itemName: String
            let itemCategory: String?
            let quantityNeeded: Int?
        }
        let body = ItemRequest(eventId: eventId, itemName: input.itemName, itemCategory: input.itemCategory, quantityNeeded: input.quantityNeeded)
        return try await api.post("items", body: body)
    }

    func claimItem(itemId: String) async throws {
        try await api.putVoid("items", queryItems: [
            .init(name: "id", value: itemId),
            .init(name: "action", value: "claim")
        ])
    }

    func unclaimItem(itemId: String) async throws {
        try await api.putVoid("items", queryItems: [
            .init(name: "id", value: itemId),
            .init(name: "action", value: "unclaim")
        ])
    }

    func addGame(input: AddEventGameInput) async throws -> EventGame {
        try await api.post("event-games", body: input)
    }

    func removeGame(gameId: String) async throws {
        try await api.deleteVoid("event-games", queryItems: [.init(name: "id", value: gameId)])
    }
}

File: android/app/src/main/java/com/sasquatsh/app/services/EventsService.kt
package com.sasquatsh.app.services

import com.sasquatsh.app.models.*
import com.sasquatsh.app.services.api.EventsApi
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class EventsService @Inject constructor(
    private val eventsApi: EventsApi,
    private val moshi: Moshi
) {

    private val summaryListType = Types.newParameterizedType(List::class.java, EventSummary::class.java)

    suspend fun getHostedEvents(): List<EventSummary> {
        val response = eventsApi.getHostedEvents()
        if (!response.isSuccessful) throw Exception("Failed to load hosted events")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter<List<EventSummary>>(summaryListType).fromJson(json) ?: emptyList()
    }

    suspend fun getRegisteredEvents(): List<EventSummary> {
        val response = eventsApi.getRegisteredEvents()
        if (!response.isSuccessful) throw Exception("Failed to load registered events")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter<List<EventSummary>>(summaryListType).fromJson(json) ?: emptyList()
    }

    suspend fun getPublicEvents(filter: EventSearchFilter): List<EventSummary> {
        val response = eventsApi.getPublicEvents(
            city = filter.city,
            state = filter.state,
            search = filter.search,
            gameCategory = filter.gameCategory?.value,
            gameSystem = filter.gameSystem?.value,
            difficulty = filter.difficulty?.value,
            dateFrom = filter.dateFrom,
            dateTo = filter.dateTo,
            nearbyZip = filter.nearbyZip,
            radiusMiles = filter.radiusMiles
        )
        if (!response.isSuccessful) throw Exception("Failed to load events")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter<List<EventSummary>>(summaryListType).fromJson(json) ?: emptyList()
    }

    suspend fun getGroupEvents(groupId: String): List<EventSummary> {
        val response = eventsApi.getGroupEvents(groupId = groupId)
        if (!response.isSuccessful) throw Exception("Failed to load group events")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter<List<EventSummary>>(summaryListType).fromJson(json) ?: emptyList()
    }

    suspend fun getEvent(id: String): Event {
        val response = eventsApi.getEvent(id)
        if (!response.isSuccessful) throw Exception("Failed to load event")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(Event::class.java).fromJson(json)
            ?: throw Exception("Failed to parse event")
    }

    suspend fun createEvent(input: CreateEventInput): Event {
        val response = eventsApi.createEvent(input)
        if (!response.isSuccessful) throw Exception("Failed to create event")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(Event::class.java).fromJson(json)
            ?: throw Exception("Failed to parse event")
    }

    suspend fun updateEvent(id: String, input: UpdateEventInput): Event {
        val response = eventsApi.updateEvent(id, input)
        if (!response.isSuccessful) throw Exception("Failed to update event")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(Event::class.java).fromJson(json)
            ?: throw Exception("Failed to parse event")
    }

    suspend fun deleteEvent(id: String) {
        val response = eventsApi.deleteEvent(id)
        if (!response.isSuccessful) throw Exception("Failed to delete event")
    }

    suspend fun registerForEvent(eventId: String) {
        val body = mapOf("eventId" to eventId)
        val response = eventsApi.registerForEvent(body)
        if (!response.isSuccessful) throw Exception("Failed to register for event")
    }

    suspend fun cancelRegistration(eventId: String) {
        val response = eventsApi.cancelRegistration(eventId)
        if (!response.isSuccessful) throw Exception("Failed to cancel registration")
    }

    suspend fun addItem(eventId: String, input: CreateEventItemInput) {
        val body = mapOf(
            "eventId" to eventId,
            "itemName" to input.itemName,
            "itemCategory" to (input.itemCategory ?: "other")
        )
        val response = eventsApi.addItem(body)
        if (!response.isSuccessful) throw Exception("Failed to add item")
    }

    suspend fun claimItem(itemId: String) {
        val response = eventsApi.claimItem(itemId)
        if (!response.isSuccessful) throw Exception("Failed to claim item")
    }

    suspend fun unclaimItem(itemId: String) {
        val response = eventsApi.unclaimItem(itemId)
        if (!response.isSuccessful) throw Exception("Failed to unclaim item")
    }

    suspend fun addGame(input: AddEventGameInput) {
        val response = eventsApi.addGame(input)
        if (!response.isSuccessful) throw Exception("Failed to add game")
    }

    suspend fun removeGame(gameId: String) {
        val response = eventsApi.removeGame(gameId)
        if (!response.isSuccessful) throw Exception("Failed to remove game")
    }
}
